File: distributed-job-manager/src/main/java/ru/fix/distributed/job/manager/strategy/RendezvousHashAssignmentStrategy.java
package ru.fix.distributed.job.manager.strategy;

import com.google.common.hash.Funnel;
import com.google.common.hash.Hashing;
import ru.fix.distributed.job.manager.model.AssignmentState;
import ru.fix.distributed.job.manager.model.JobId;
import ru.fix.distributed.job.manager.model.WorkItem;
import ru.fix.distributed.job.manager.model.WorkerId;
import ru.fix.distributed.job.manager.util.RendezvousHash;

import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;

public class RendezvousHashAssignmentStrategy extends AbstractAssignmentStrategy {

    @SuppressWarnings("UnstableApiUsage")
    @Override
    public AssignmentState reassignAndBalance(
            Map<JobId, Set<WorkerId>> availability,
            AssignmentState prevAssignment,
            AssignmentState currentAssignment,
            Set<WorkItem> itemsToAssign
    ) {
        final Funnel<String> stringFunnel = (from, into) -> into.putBytes(from.getBytes(StandardCharsets.UTF_8));

        for (Map.Entry<JobId, Set<WorkerId>> jobEntry : availability.entrySet()) {
            final RendezvousHash<String, String> hash = new RendezvousHash<>(
                    Hashing.murmur3_128(), stringFunnel, stringFunnel, new ArrayList<>()
            );
            jobEntry.getValue().forEach(worker -> {
                hash.add(worker.getId());
                currentAssignment.putIfAbsent(worker, new HashSet<>());
            });

            Set<WorkItem> availableItems = getWorkItemsByJob(jobEntry.getKey(), itemsToAssign);
            for (WorkItem item : availableItems) {
                if (currentAssignment.containsWorkItem(item)) {
                    continue;
                }

                String workerId = hash.get(item.getJobId().getId() + ":" + item.getId());
                currentAssignment.addWorkItem(new WorkerId(workerId), item);
                itemsToAssign.remove(item);
            }
        }

        return currentAssignment;
    }
}

File: distributed-job-manager/src/main/kotlin/ru/fix/distributed/job/manager/strategy/RendezvousHashAssignmentStrategy.kt
package ru.fix.distributed.job.manager.strategy

import com.google.common.hash.Funnel
import com.google.common.hash.Hashing
import ru.fix.distributed.job.manager.model.AssignmentState
import ru.fix.distributed.job.manager.model.JobId
import ru.fix.distributed.job.manager.model.WorkItem
import ru.fix.distributed.job.manager.model.WorkerId
import ru.fix.distributed.job.manager.util.RendezvousHash
import java.nio.charset.StandardCharsets
import java.util.*
import kotlin.collections.ArrayList

class RendezvousHashAssignmentStrategy : AbstractAssignmentStrategy() {

    override fun reassignAndBalance(
            availability: MutableMap<JobId, MutableSet<WorkerId>>,
            prevAssignment: AssignmentState,
            currentAssignment: AssignmentState,
            itemsToAssign: MutableSet<WorkItem>
    ): AssignmentState {
        val stringFunnel = Funnel<String> { from, into ->
            into.putBytes(from.toByteArray(StandardCharsets.UTF_8))
        }

        for ((jobId, availableWorkers) in availability) {
            val hash = RendezvousHash<String, String>(
                    Hashing.murmur3_128(), stringFunnel, stringFunnel, ArrayList()
            )
            availableWorkers.forEach { worker ->
                hash.add(worker.id)
                currentAssignment.putIfAbsent(worker, HashSet<WorkItem>())
            }

            val availableItems = getWorkItemsByJob(jobId, itemsToAssign)
            for (item in availableItems) {
                if (currentAssignment.containsWorkItem(item)) {
                    continue
                }

                val workerId = hash.get(item.jobId.id + ":" + item.id)
                currentAssignment.addWorkItem(WorkerId(workerId), item)
                itemsToAssign.remove(item)
            }
        }
        return currentAssignment
    }
}
